content/blog/2023-04-01-key-expressions-1.md

```diff
@@ -0,0 +1,147 @@
+    For example, `*/**/*`, `*/**/**/*` and `**/*/*/**` would mean the same thing (the set of all keys that are made of at least 2 chunks), but by repeatedly applying the `**/* -> */**` and `**/** -> **` rules, they all come down to the same `*/*/**`.
+
+With steps 1 and 2, ambiguities in the language disappear. With step 3, we gain extensibility for future features, which will one day allow us to express more precise sets than currently possible. But step 4 is the true hero of the story: thanks to bijectivity, there's no longer a need to worry about different strings meaning the same thing, which may have trapped many people that haven't spent the last year obsessing over KEs. Bijectivity also greatly simplifies the implementation of data structures tailor-made for KEs, such as the one we'll explore in the next part.
+
+# The KE Tree: a Zenoh flavoured data structure.
+
+One thing you might have noticed is that with all these slashes, KEs definitely take after paths. One other aspect they take from paths is their intrisic hierarchical nature: like most good address spaces, KEs are hierarchical.
+
+And what data-structure rimes with hierarchy? The section title spoiled it, it's the _tree_. But the KeTree isn't just any old tree: it's a tree that's made to help you treat KEs as the sets they represent, complete with _intersection_ and _inclusion_ comparisons.
+
+## Wait, what does it mean for KEs to _intersect_ and _include_?
+
+Glad you asked, inner monologue of my future reader, forcing you to read this question in your mind was definitely helpful!
+
+As aluded to repeatedly in this post, KEs define sets of keys, and operations in Zenoh are addressed by KE, so they affect sets of things. In fact, there are four ways an operation on a KE can affect data associated with a given KE:
+1. If the KEs define disjoint sets (where there doesn't exist a key that belongs to both), they just don't interract.
+2. If the operation KE and the existing data's KE define sets that have keys in common, like `a/*` and `*/b` both contain the `a/b` key, they are said to _intersect_. That means the operation will affect at least a subset of the existing data. Intersection is always symmetric: A intersects with B implies that B intersects with A.
+3. If the operation KE's set contains all keys defined by the existing data's KE, the operation KE is said to _include_ the existing data KE. This means that the operation will affect _all_ of the existing data. Note that inclusion is generally asymmetric, but that "A includes B" or "B includes A" implies that "A and B intersect".
```

Comment:
"A includes B" or "B includes A" implies that "A and B intersect" -> doesn't it mean that A = B?

Reply:
Since you are addressing A=B in the next line, can we just remove that part and just say "Note that inclusion is generally asymmetric."? Point 4, will define the special case where it is symmetric.

Reply:
Sorry, I misread OR for AND, my bad!